src/kakarot/instructions/system_operations.cairo

```diff
@@ -301,6 +301,60 @@ namespace SystemOperations {
 
         return sub_ctx;
     }
+
+    // @notice SELFDESTRUCT operation.
+    // @dev
+    // @custom:since Frontier
+    // @custom:group System Operations
+    // @custom:gas 3000 + dynamic gas
+    // @custom:stack_consumed_elements 1
+    // @return The pointer to the updated execution_context.
+    func exec_selfdestruct{
+        syscall_ptr: felt*,
+        pedersen_ptr: HashBuiltin*,
+        range_check_ptr,
+        bitwise_ptr: BitwiseBuiltin*,
+    }(ctx: model.ExecutionContext*) -> model.ExecutionContext* {
+        alloc_locals;
+
+        // Get stack and memory from context
+        let stack = ctx.stack;
+
+        // Stack input:
+        // 0 - address: account to send the current balance to
+        let (stack, address_uint256) = Stack.pop(stack);
+        let address_felt = Helpers.uint256_to_felt(address_uint256);
+        let ctx = ExecutionContext.update_stack(ctx, stack);
+        // Get the number of native tokens owned by the given starknet
+        // account and transfer them to receiver
+        let (native_token_address_) = native_token_address.read();
+        let (balance: Uint256) = IEth.balanceOf(
+            contract_address=native_token_address_, account=ctx.starknet_contract_address
+        );
+        let (success) = IEth.transfer(contract_address=native_token_address_, recipient=address_felt, amount=balance);
+        with_attr error_message("Kakarot: Transfer failed") {
+            assert success = TRUE;
+        }
+
+        // Save contract to be destroyed at the end of the transaction in a NEW array
+        // Directly filling ctx.destroy_contract produces error "Expected memory address to be relocatable value. Found: 0."
```

Comment:
If we always create a new array, how can we destruct several contracts in the same tx? (My understanding is that it’s possible that several sub context call SELFDESTRUCT and at the end of the tx there are consequently several contracts to be destroyed)

Reply:
Yes you're right. At first I was updating the current array but it was raising this error _"Expected memory address to be relocatable value. Found: 0."_ and I don't really know why but creating a new array solved this issue. Also I'm filling the newly created array with the old one data with Helpers.fill_array() so contract to destroy collected from other contexts are not lost.

Reply:
I am quite surprised because for `return_data` for example we don't create a new array but fill with `Memory.n_load` the originally empty array. AFAIK in cairo an "array" is just an initialized segment, so this error may indicate that the pointers were not used correctly. Can you double check?

Reply:
My bad, I had not initialized destroy_contract array in my test